applications/StructuralMechanicsApplication/python_scripts/structural_response.py

```diff
@@ -220,7 +220,215 @@ def GetShapeGradient(self):
+        print("\n> Starting primal analysis for response:", self.identifier)
+        startTime = timer.time()
+        self.primal_analysis.time = self.primal_analysis._GetSolver().AdvanceInTime(self.primal_analysis.time)
+        self.primal_analysis.InitializeSolutionStep()
+        self.primal_analysis._GetSolver().Predict()
+        self.primal_analysis._GetSolver().SolveSolutionStep()
+        self.primal_analysis.FinalizeSolutionStep()
+        self.primal_analysis.OutputSolutionStep()
+        print("> Time needed for solving the primal analysis = ",round(timer.time() - startTime,2),"s")
+
+        # TODO the response value calculation for stresses currently only works on the adjoint modelpart
+        # this needs to be improved, also the response value should be calculated on the PRIMAL modelpart!!
+        self.adjoint_analysis.time = self.adjoint_analysis._GetSolver().AdvanceInTime(self.adjoint_analysis.time)
+        self.adjoint_analysis.InitializeSolutionStep()
+
+    def CalculateValue(self):
+        startTime = timer.time()
+        value = self.response_function_utility.CalculateValue(self.adjoint_model_part)
+        print("> Time needed for calculating the response value = ",round(timer.time() - startTime,2),"s")
+
+        self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.RESPONSE_VALUE] = value
+
+
+    def CalculateGradient(self):
+        print("\n> Starting adjoint analysis for response:", self.identifier)
+        startTime = timer.time()
+        self.adjoint_analysis._GetSolver().Predict()
+        self.adjoint_analysis._GetSolver().SolveSolutionStep()
+        self.adjoint_analysis.FinalizeSolutionStep()
+        self.adjoint_analysis.OutputSolutionStep()
+        print("> Time needed for solving the adjoint analysis = ",round(timer.time() - startTime,2),"s")
+
+
+    def GetValue(self):
+        return self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.RESPONSE_VALUE]
+
+
+    def GetShapeGradient(self):
+        gradient = {}
+        for node in self.adjoint_model_part.Nodes:
+            gradient[node.Id] = node.GetSolutionStepValue(SHAPE_SENSITIVITY)
+        return gradient
+
+
+    def FinalizeSolutionStep(self):
+        # TODO reset DISPLACEMENT, ROTATION ADJOINT_DISPLACEMENT and ADJOINT_ROTATION
+        pass
+
+
+    def Finalize(self):
+        self.primal_analysis.Finalize()
+        self.adjoint_analysis.Finalize()
+
+# ==============================================================================
+class AdjointStrainEnergyResponse(ResponseFunctionBase):
+    """Linear static adjoint response function.
+    - runs the primal analysis (writes the primal results to an .h5 file)
+    - reads the primal results from the .h5 file into the adjoint model part
+    - uses primal results to calculate value
+    - uses primal results to calculate gradient
+
+    Note: because of the special property of the adjoint strain energy response,
```

Comment:
Doesn't this only work for linear problems? Should we rename the class to reflect this?

Reply:
Done